macros/src/serializable.rs

```diff
@@ -37,43 +38,94 @@ pub(crate) fn impl_derive_serializable(item: TokenStream) -> TokenStream {
         _ => HashSet::default(),
     };
 
-    let collect_types = if field_types.is_empty() {
-        quote! { types.entry(Self::ident()).or_insert_with(Self::ty); }
-    } else {
-        let field_types = field_types.iter();
-        let generic_params = generics.type_params();
-        quote! {
-            if let std::collections::btree_map::Entry::Vacant(entry) = types.entry(Self::ident()) {
-                entry.insert(Self::ty());
-                #( #field_types::collect_types(types); )*
-            }
+    // Remove any bounds from the generic types and store them separately.
+    // Otherwise, collect_types will be called like `Foo::<T: MyTrait>::collect_types()` and where clauses
+    // will be incorrect, too.
+    let mut bounds = BTreeMap::new();
+    for param in generics.type_params_mut() {
+        // For every parameter we want to either extract the existing trait bounds, or, if there
+        // were no existing bounds, we will mark the parameter as having no bounds.
 
-            #( #generic_params::collect_types(types); )*
-        }
-    };
+        param.bounds = if param.bounds.is_empty() {
+            // No existing bound found, so mark this parameter as having 'None' as a bound
+            bounds.insert(param.ident.clone(), vec![]);
+            Punctuated::new()
+        } else {
+            param
+                .clone()
+                .bounds
+                .into_iter()
+                .filter(|bound| {
+                    match &bound {
+                        TypeParamBound::Trait(tr) => {
+                            // Extract the bound and remove it from the param
+                            bounds
+                                .entry(param.ident.clone())
+                                .or_insert_with(Vec::new)
+                                .push(tr.clone());
+                            false
+                        }
+                        // Ignore other bound types (lifetimes) for now
+                        _ => true,
+                    }
+                })
```

Comment:
I think I would have used [partition()](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.partition) for filtering the bounds, and then applied the side-effects to the extracted bounds.

Reply:
Good tip, I don't think I was even aware of `partition`. Would have worked well here, indeed.